File: analysis/optimize_gradient_weights.py
#!/usr/bin/env python3
"""
optimize_gradient_weights.py
Piecewise-linear (gradient) EF weight optimization for Potomac Pulse.

Replaces hard step-function cutoffs with smooth interpolation between anchor
points. Uses iterative coordinate descent + refinement to find optimal weights
at each anchor, enforcing monotonicity throughout.

Data: ef_lf_daily_longterm.csv — 5,220 deduplicated daily obs (2011-2026)
Model: ensemble = (1-w)*yesterday_lf + w * (126 * ef_stage^2.46)
"""

import numpy as np
import pandas as pd

# ── Configuration ──────────────────────────────────────────────────────────
ANCHOR_FLOWS = np.array([0, 3000, 6000, 10000, 15000, 25000, 50000], dtype=float)
INITIAL_WEIGHTS = np.array([0.1, 0.1, 0.1, 0.2, 0.7, 0.7, 0.7])
W_MIN, W_MAX = 0.0, 0.80
COARSE_STEP = 0.05
FINE_RADIUS = 0.05
FINE_STEP = 0.01
N_PASSES = 5

# Current step-function weights for comparison
STEP_THRESHOLDS = [3000, 6000, 15000]
STEP_WEIGHTS = [0.10, 0.10, 0.20, 0.70]  # <3k, 3-6k, 6-15k, >15k

DATA_PATH = "/Users/sebjilke/Desktop/PotomacPulse/analysis/ef_lf_daily_longterm.csv"
OUT_CSV = "/Users/sebjilke/Desktop/PotomacPulse/analysis/gradient_weights_python.csv"

# ── Flow regime bins for RMSE breakdown ────────────────────────────────────
REGIME_BINS = [
    ("<3k", 0, 3000),
    ("3-6k", 3000, 6000),
    ("6-10k", 6000, 10000),
    ("10-15k", 10000, 15000),
    ("15-25k", 15000, 25000),
    ("25-50k", 25000, 50000),
    (">50k", 50000, np.inf),
]


def load_and_prepare():
    """Load CSV, deduplicate, sort, compute EF predicted, build consecutive pairs."""
    df = pd.read_csv(DATA_PATH, parse_dates=["date"])
    df = df.drop_duplicates(subset="date", keep="first").sort_values("date").reset_index(drop=True)
    print(f"Loaded {len(df)} unique daily observations")
    print(f"Date range: {df['date'].min().date()} to {df['date'].max().date()}")

    # EF power-law prediction
    df["ef_predicted"] = 126.0 * df["ef_stage"] ** 2.46

    # Build consecutive-day pairs
    dates = df["date"].values
    ef_pred = df["ef_predicted"].values
    lf_q = df["lf_discharge"].values

    actual_list = []
    yesterday_list = []
    ef_p_list = []

    for i in range(1, len(df)):
        gap = (dates[i] - dates[i - 1]) / np.timedelta64(1, "D")
        if gap == 1.0:
            actual_list.append(lf_q[i])
            yesterday_list.append(lf_q[i - 1])
            ef_p_list.append(ef_pred[i])

    actual = np.array(actual_list)
    yesterday = np.array(yesterday_list)
    ef_p = np.array(ef_p_list)

    print(f"Consecutive-day pairs: {len(actual)}")
    return actual, yesterday, ef_p


def interpolate_weight(flow, anchor_flows, anchor_weights):
    """Piecewise-linear interpolation of weight at a given flow level."""
    if flow <= anchor_flows[0]:
        return anchor_weights[0]
    if flow >= anchor_flows[-1]:
        return anchor_weights[-1]
    idx = np.searchsorted(anchor_flows, flow, side="right") - 1
    f0, f1 = anchor_flows[idx], anchor_flows[idx + 1]
    w0, w1 = anchor_weights[idx], anchor_weights[idx + 1]
    t = (flow - f0) / (f1 - f0)
    return w0 + (w1 - w0) * t


def interpolate_weights_vectorized(flows, anchor_flows, anchor_weights):
    """Vectorized piecewise-linear interpolation for an array of flows."""
    weights = np.empty(len(flows), dtype=float)
    for i in range(len(flows)):
        weights[i] = interpolate_weight(flows[i], anchor_flows, anchor_weights)
    return weights


def compute_rmse(actual, yesterday, ef_pred, weights_vec):
    """Compute RMSE of blended ensemble given per-observation weights."""
    ensemble = (1.0 - weights_vec) * yesterday + weights_vec * ef_pred
    return np.sqrt(np.mean((ensemble - actual) ** 2))


def compute_rmse_for_anchors(actual, yesterday, ef_pred, anchor_flows, anchor_weights):
    """Compute overall RMSE for a given set of anchor weights."""
    w_vec = interpolate_weights_vectorized(actual, anchor_flows, anchor_weights)
    return compute_rmse(actual, yesterday, ef_pred, w_vec)


def enforce_monotonicity(weights):
    """Enforce non-decreasing constraint (returns new array)."""
    w = weights.copy()
    for i in range(1, len(w)):
        if w[i] < w[i - 1]:
            w[i] = w[i - 1]
    return w


def step_function_weight(flow):
    """Current step-function weight for comparison."""
    if flow < STEP_THRESHOLDS[0]:
        return STEP_WEIGHTS[0]
    elif flow < STEP_THRESHOLDS[1]:
        return STEP_WEIGHTS[1]
    elif flow < STEP_THRESHOLDS[2]:
        return STEP_WEIGHTS[2]
    else:
        return STEP_WEIGHTS[3]


def rmse_by_regime(actual, yesterday, ef_pred, weight_func):
    """Compute RMSE breakdown by flow regime."""
    results = []
    for label, lo, hi in REGIME_BINS:
        mask = (actual >= lo) & (actual < hi)
        n = mask.sum()
        if n == 0:
            results.append((label, n, np.nan))
            continue
        w_vec = np.array([weight_func(f) for f in actual[mask]])
        ensemble = (1.0 - w_vec) * yesterday[mask] + w_vec * ef_pred[mask]
        rmse = np.sqrt(np.mean((ensemble - actual[mask]) ** 2))
        results.append((label, n, rmse))
    return results


def optimize():
    """Main optimization routine."""
    actual, yesterday, ef_pred = load_and_prepare()

    # ── Step-function baseline ─────────────────────────────────────────────
    step_w = np.array([step_function_weight(f) for f in actual])
    step_rmse = compute_rmse(actual, yesterday, ef_pred, step_w)
    print(f"\n{'='*70}")
    print(f"CURRENT STEP-FUNCTION BASELINE")
    print(f"  Weights: <3k=0.10, 3-6k=0.10, 6-15k=0.20, >15k=0.70")
    print(f"  Overall RMSE: {step_rmse:.1f} cfs")

    # ── Coordinate descent — coarse ───────────────────────────────────────
    print(f"\n{'='*70}")
    print("COORDINATE DESCENT — COARSE (step=0.05, 5 passes)")
    weights = INITIAL_WEIGHTS.copy()
    coarse_grid = np.arange(W_MIN, W_MAX + COARSE_STEP / 2, COARSE_STEP)

    for pass_num in range(1, N_PASSES + 1):
        for a in range(len(ANCHOR_FLOWS)):
            best_w = weights[a]
            best_rmse = np.inf
            for candidate in coarse_grid:
                trial = weights.copy()
                trial[a] = candidate
                trial = enforce_monotonicity(trial)
                rmse = compute_rmse_for_anchors(actual, yesterday, ef_pred, ANCHOR_FLOWS, trial)
                if rmse < best_rmse:
                    best_rmse = rmse
                    best_w = candidate
            weights[a] = best_w
            weights = enforce_monotonicity(weights)
        rmse_now = compute_rmse_for_anchors(actual, yesterday, ef_pred, ANCHOR_FLOWS, weights)
        print(f"  Pass {pass_num}: RMSE={rmse_now:.1f}  weights={np.round(weights, 2)}")

    # ── Coordinate descent — fine ─────────────────────────────────────────
    print(f"\nCOORDINATE DESCENT — FINE (step=0.01, +/-0.05)")
    for pass_num in range(1, 4):
        for a in range(len(ANCHOR_FLOWS)):
            lo = max(W_MIN, weights[a] - FINE_RADIUS)
            hi = min(W_MAX, weights[a] + FINE_RADIUS)
            fine_grid = np.arange(lo, hi + FINE_STEP / 2, FINE_STEP)
            best_w = weights[a]
            best_rmse = np.inf
            for candidate in fine_grid:
                trial = weights.copy()
                trial[a] = round(candidate, 4)
                trial = enforce_monotonicity(trial)
                rmse = compute_rmse_for_anchors(actual, yesterday, ef_pred, ANCHOR_FLOWS, trial)
                if rmse < best_rmse:
                    best_rmse = rmse
                    best_w = round(candidate, 4)
            weights[a] = best_w
            weights = enforce_monotonicity(weights)
        rmse_now = compute_rmse_for_anchors(actual, yesterday, ef_pred, ANCHOR_FLOWS, weights)
        print(f"  Fine pass {pass_num}: RMSE={rmse_now:.1f}  weights={np.round(weights, 2)}")

    # ── Round to 1 decimal ────────────────────────────────────────────────
    weights_rounded = np.round(weights, 1)
    weights_rounded = enforce_monotonicity(weights_rounded)
    rmse_rounded = compute_rmse_for_anchors(actual, yesterday, ef_pred, ANCHOR_FLOWS, weights_rounded)

    # ── Final results ─────────────────────────────────────────────────────
    print(f"\n{'='*70}")
    print("FINAL OPTIMIZED GRADIENT WEIGHTS (rounded to 0.1)")
    print(f"{'='*70}")
    print(f"{'Anchor Flow (cfs)':>20s} {'Weight':>10s}")
    print(f"{'-'*32}")
    for f, w in zip(ANCHOR_FLOWS, weights_rounded):
        print(f"{f:>20,.0f} {w:>10.1f}")

    print(f"\n  Gradient RMSE (rounded):  {rmse_rounded:.1f} cfs")
    print(f"  Step-function RMSE:       {step_rmse:.1f} cfs")
    diff = step_rmse - rmse_rounded
    pct = diff / step_rmse * 100
    if diff > 0:
        print(f"  Improvement:              {diff:.1f} cfs ({pct:.1f}% reduction)")
    else:
        print(f"  Difference:               {-diff:.1f} cfs worse ({-pct:.1f}%)")

    # ── RMSE by flow regime ───────────────────────────────────────────────
    def gradient_weight_func(flow):
        return interpolate_weight(flow, ANCHOR_FLOWS, weights_rounded)

    print(f"\n{'='*70}")
    print("RMSE BY FLOW REGIME")
    print(f"{'='*70}")
    print(f"{'Regime':>10s} {'N pairs':>10s} {'Step RMSE':>12s} {'Gradient RMSE':>15s} {'Change':>10s}")
    print(f"{'-'*60}")

    regime_step = rmse_by_regime(actual, yesterday, ef_pred, step_function_weight)
    regime_grad = rmse_by_regime(actual, yesterday, ef_pred, gradient_weight_func)

    for (label, n_s, rmse_s), (_, n_g, rmse_g) in zip(regime_step, regime_grad):
        if n_s == 0:
            print(f"{label:>10s} {n_s:>10d} {'n/a':>12s} {'n/a':>15s} {'':>10s}")
        else:
            chg = rmse_g - rmse_s
            sign = "+" if chg > 0 else ""
            print(f"{label:>10s} {n_s:>10d} {rmse_s:>12.1f} {rmse_g:>15.1f} {sign}{chg:>9.1f}")

    # ── Per-anchor local stats (for CSV) ──────────────────────────────────
    anchor_stats = []
    for idx_a, (af, aw) in enumerate(zip(ANCHOR_FLOWS, weights_rounded)):
        if idx_a == 0:
            lo_bound = 0
        else:
            lo_bound = (ANCHOR_FLOWS[idx_a - 1] + af) / 2
        if idx_a == len(ANCHOR_FLOWS) - 1:
            hi_bound = np.inf
        else:
            hi_bound = (af + ANCHOR_FLOWS[idx_a + 1]) / 2

        mask = (actual >= lo_bound) & (actual < hi_bound)
        n_nearby = mask.sum()
        if n_nearby > 0:
            w_local = np.full(n_nearby, aw)
            ens = (1.0 - w_local) * yesterday[mask] + w_local * ef_pred[mask]
            rmse_local = np.sqrt(np.mean((ens - actual[mask]) ** 2))
        else:
            rmse_local = np.nan
        anchor_stats.append((af, aw, n_nearby, round(rmse_local, 1) if not np.isnan(rmse_local) else np.nan))

    # ── Save CSV ──────────────────────────────────────────────────────────
    out_df = pd.DataFrame(anchor_stats, columns=["anchor_flow", "optimal_weight", "n_obs_nearby", "rmse_local"])
    out_df.to_csv(OUT_CSV, index=False)
    print(f"\nSaved: {OUT_CSV}")

    # ── Print for code integration ────────────────────────────────────────
    w_str = ", ".join([f"{w:.1f}" for w in weights_rounded])
    a_str = ", ".join([f"{int(a)}" for a in ANCHOR_FLOWS])
    print(f"\nFor code:")
    print(f"  Anchors: [{a_str}]")
    print(f"  Weights: [{w_str}]")

    return weights_rounded, rmse_rounded


if __name__ == "__main__":
    optimize()

File: analysis/optimize_gradient_weights.R
#!/usr/bin/env Rscript
# =============================================================================
# optimize_gradient_weights.R
# Piecewise-linear (gradient) EF weight optimization via coordinate descent
# Potomac Pulse — v26.0
# =============================================================================

cat("=============================================================
")
cat("  Gradient EF Weight Optimization (Piecewise-Linear)
")
cat("  Potomac Pulse — Coordinate Descent Method
")
cat("=============================================================

")

# --- 1. Load and prepare data ------------------------------------------------
data_path <- "/Users/sebjilke/Desktop/PotomacPulse/analysis/ef_lf_daily_longterm.csv"
df <- read.csv(data_path, stringsAsFactors = FALSE)
cat(sprintf("Raw data loaded: %d rows
", nrow(df)))

# Parse dates
df$date <- as.Date(df$date)

# Deduplicate by date (keep first occurrence)
df <- df[!duplicated(df$date), ]
df <- df[order(df$date), ]
cat(sprintf("After dedup by date: %d rows
", nrow(df)))

# Remove NA rows
df <- df[complete.cases(df), ]
cat(sprintf("After removing NAs: %d rows
", nrow(df)))

# --- 2. Compute EF predicted -------------------------------------------------
df$ef_predicted <- 126 * df$ef_stage^2.46
cat(sprintf("EF predicted range: %.0f - %.0f cfs
",
            min(df$ef_predicted), max(df$ef_predicted)))
cat(sprintf("LF discharge range: %.0f - %.0f cfs

",
            min(df$lf_discharge), max(df$lf_discharge)))

# --- 3. Build consecutive-day pairs ------------------------------------------
n <- nrow(df)
day_gap <- as.numeric(diff(df$date))
consec <- which(day_gap == 1)

pairs <- data.frame(
  actual       = df$lf_discharge[consec + 1],
  yesterday_lf = df$lf_discharge[consec],
  ef_pred      = df$ef_predicted[consec + 1],
  flow_level   = df$lf_discharge[consec + 1]
)

cat(sprintf("Consecutive-day pairs: %d

", nrow(pairs)))

# --- 4. Define anchor points and helper functions -----------------------------
anchors <- c(0, 3000, 6000, 10000, 15000, 25000, 50000)
n_anchors <- length(anchors)

interpolate_weight_vec <- function(flow, anchor_flows, anchor_weights) {
  n_a <- length(anchor_flows)
  w <- rep(anchor_weights[1], length(flow))
  above <- flow >= anchor_flows[n_a]
  w[above] <- anchor_weights[n_a]
  for (j in 1:(n_a - 1)) {
    in_range <- flow > anchor_flows[j] & flow < anchor_flows[j + 1]
    if (any(in_range)) {
      frac <- (flow[in_range] - anchor_flows[j]) /
              (anchor_flows[j + 1] - anchor_flows[j])
      w[in_range] <- anchor_weights[j] +
                     (anchor_weights[j + 1] - anchor_weights[j]) * frac
    }
    at_anchor <- flow == anchor_flows[j + 1]
    w[at_anchor] <- anchor_weights[j + 1]
  }
  return(w)
}

compute_rmse <- function(anchor_weights, pairs_df, anchor_flows) {
  w <- interpolate_weight_vec(pairs_df$flow_level, anchor_flows, anchor_weights)
  ensemble <- (1 - w) * pairs_df$yesterday_lf + w * pairs_df$ef_pred
  rmse <- sqrt(mean((ensemble - pairs_df$actual)^2))
  return(rmse)
}

enforce_monotonic <- function(weights) {
  for (i in 2:length(weights)) {
    if (weights[i] < weights[i - 1]) {
      weights[i] <- weights[i - 1]
    }
  }
  return(weights)
}

# --- 5. Current step-function RMSE (baseline) --------------------------------
cat("--- Current Step-Function Baseline ---
")
step_weights_at_obs <- function(flow) {
  ifelse(flow < 3000, 0.1,
    ifelse(flow < 6000, 0.1,
      ifelse(flow < 15000, 0.2,
        0.7)))
}
w_step <- step_weights_at_obs(pairs$flow_level)
ensemble_step <- (1 - w_step) * pairs$yesterday_lf + w_step * pairs$ef_pred
rmse_step <- sqrt(mean((ensemble_step - pairs$actual)^2))
cat(sprintf("Step-function overall RMSE: %.1f cfs

", rmse_step))

# --- 6. Coordinate descent optimization (coarse) -----------------------------
cat("--- Coordinate Descent Optimization (Coarse: step 0.05) ---
")
current_weights <- c(0.1, 0.1, 0.1, 0.2, 0.7, 0.7, 0.7)
coarse_grid <- seq(0.00, 0.80, by = 0.05)
best_rmse <- compute_rmse(current_weights, pairs, anchors)
cat(sprintf("Initial RMSE: %.1f cfs
", best_rmse))

for (pass in 1:5) {
  improved <- FALSE
  for (a in 1:n_anchors) {
    best_w_for_anchor <- current_weights[a]
    best_rmse_for_anchor <- best_rmse
    for (w_candidate in coarse_grid) {
      trial_weights <- current_weights
      trial_weights[a] <- w_candidate
      trial_weights <- enforce_monotonic(trial_weights)
      trial_rmse <- compute_rmse(trial_weights, pairs, anchors)
      if (trial_rmse < best_rmse_for_anchor) {
        best_rmse_for_anchor <- trial_rmse
        best_w_for_anchor <- w_candidate
      }
    }
    if (best_rmse_for_anchor < best_rmse) {
      current_weights[a] <- best_w_for_anchor
      current_weights <- enforce_monotonic(current_weights)
      best_rmse <- best_rmse_for_anchor
      improved <- TRUE
    }
  }
  cat(sprintf("  Pass %d: RMSE = %.2f cfs | Weights = [%s]
",
              pass, best_rmse,
              paste(sprintf("%.2f", current_weights), collapse = ", ")))
  if (!improved) {
    cat("  Converged (no improvement).
")
    break
  }
}
cat(sprintf("
Coarse result RMSE: %.2f cfs
", best_rmse))
cat(sprintf("Coarse weights: [%s]

",
            paste(sprintf("%.2f", current_weights), collapse = ", ")))

# --- 7. Fine refinement (step 0.01) ------------------------------------------
cat("--- Fine Refinement (step 0.01, +/- 0.05 around coarse) ---
")
for (pass in 1:5) {
  improved <- FALSE
  for (a in 1:n_anchors) {
    lo <- max(0.00, current_weights[a] - 0.05)
    hi <- min(0.80, current_weights[a] + 0.05)
    fine_grid <- seq(lo, hi, by = 0.01)
    best_w_for_anchor <- current_weights[a]
    best_rmse_for_anchor <- best_rmse
    for (w_candidate in fine_grid) {
      trial_weights <- current_weights
      trial_weights[a] <- w_candidate
      trial_weights <- enforce_monotonic(trial_weights)
      trial_rmse <- compute_rmse(trial_weights, pairs, anchors)
      if (trial_rmse < best_rmse_for_anchor) {
        best_rmse_for_anchor <- trial_rmse
        best_w_for_anchor <- w_candidate
      }
    }
    if (best_rmse_for_anchor < best_rmse) {
      current_weights[a] <- best_w_for_anchor
      current_weights <- enforce_monotonic(current_weights)
      best_rmse <- best_rmse_for_anchor
      improved <- TRUE
    }
  }
  cat(sprintf("  Refine pass %d: RMSE = %.4f cfs | Weights = [%s]
",
              pass, best_rmse,
              paste(sprintf("%.2f", current_weights), collapse = ", ")))
  if (!improved) {
    cat("  Converged (no improvement).
")
    break
  }
}

# --- 8. Round to 1 decimal (0.1 precision) -----------------------------------
cat("
--- Rounding to 0.1 precision ---
")
rounded_weights <- round(current_weights, 1)
rounded_weights <- enforce_monotonic(rounded_weights)
rmse_rounded <- compute_rmse(rounded_weights, pairs, anchors)
cat(sprintf("Rounded weights: [%s]
",
            paste(sprintf("%.1f", rounded_weights), collapse = ", ")))
cat(sprintf("Rounded RMSE: %.2f cfs

", rmse_rounded))

# --- 9. Detailed results table ------------------------------------------------
cat("=============================================================
")
cat("  RESULTS: Optimal Piecewise-Linear (Gradient) EF Weights
")
cat("=============================================================

")
regime_labels <- c("< 3k", "3k-6k", "6k-10k", "10k-15k", "15k-25k", "25k-50k", "> 50k")
regime_lo <- c(0, 3000, 6000, 10000, 15000, 25000, 50000)
regime_hi <- c(3000, 6000, 10000, 15000, 25000, 50000, Inf)
cat(sprintf("%-12s | %-8s | %-8s | %-12s | %-12s | %-12s
",
            "Flow Regime", "Wt(mid)", "N obs", "RMSE Grad.", "RMSE Step", "Improvement"))
cat(paste(rep("-", 78), collapse = ""), "
")

for (r in seq_along(regime_labels)) {
  in_regime <- pairs$flow_level >= regime_lo[r] & pairs$flow_level < regime_hi[r]
  n_regime <- sum(in_regime)
  if (n_regime > 0) {
    w_grad <- interpolate_weight_vec(pairs$flow_level[in_regime], anchors, rounded_weights)
    ens_grad <- (1 - w_grad) * pairs$yesterday_lf[in_regime] +
                w_grad * pairs$ef_pred[in_regime]
    rmse_grad_r <- sqrt(mean((ens_grad - pairs$actual[in_regime])^2))
    w_s <- step_weights_at_obs(pairs$flow_level[in_regime])
    ens_s <- (1 - w_s) * pairs$yesterday_lf[in_regime] +
             w_s * pairs$ef_pred[in_regime]
    rmse_step_r <- sqrt(mean((ens_s - pairs$actual[in_regime])^2))
    pct_change <- (rmse_grad_r - rmse_step_r) / rmse_step_r * 100
    mid_flow <- (regime_lo[r] + min(regime_hi[r], 60000)) / 2
    mid_w <- interpolate_weight_vec(mid_flow, anchors, rounded_weights)
    cat(sprintf("%-12s | %-8.2f | %-8d | %-12.1f | %-12.1f | %+.1f%%
",
                regime_labels[r], mid_w, n_regime,
                rmse_grad_r, rmse_step_r, pct_change))
  } else {
    cat(sprintf("%-12s | %-8s | %-8d | %-12s | %-12s | %s
",
                regime_labels[r], "N/A", 0, "N/A", "N/A", "N/A"))
  }
}
cat(paste(rep("-", 78), collapse = ""), "
")
cat(sprintf("
%-12s | %-8s | %-8d | %-12.1f | %-12.1f | %+.1f%%
",
            "OVERALL", "", nrow(pairs), rmse_rounded, rmse_step,
            (rmse_rounded - rmse_step) / rmse_step * 100))
cat(sprintf("
Total consecutive-day pairs used: %d
", nrow(pairs)))

# --- 10. Weight comparison table ----------------------------------------------
cat("
=============================================================
")
cat("  Weight Comparison: Step vs Gradient
")
cat("=============================================================

")
cat(sprintf("%-14s | %-14s | %-14s
",
            "Anchor (cfs)", "Step Weight", "Gradient Weight"))
cat(paste(rep("-", 48), collapse = ""), "
")
step_anchor_weights <- c(0.1, 0.1, 0.1, 0.2, 0.7, 0.7, 0.7)
for (a in 1:n_anchors) {
  cat(sprintf("%-14s | %-14.1f | %-14.1f
",
              formatC(anchors[a], format = "d", big.mark = ","),
              step_anchor_weights[a], rounded_weights[a]))
}

# --- 11. Save CSV output ------------------------------------------------------
out_path <- "/Users/sebjilke/Desktop/PotomacPulse/analysis/gradient_weights_R.csv"
local_rmse <- numeric(n_anchors)
local_n <- numeric(n_anchors)
for (a in 1:n_anchors) {
  if (a == 1) {
    lo_bound <- 0
  } else {
    lo_bound <- (anchors[a] + anchors[a - 1]) / 2
  }
  if (a == n_anchors) {
    hi_bound <- Inf
  } else {
    hi_bound <- (anchors[a] + anchors[a + 1]) / 2
  }
  in_band <- pairs$flow_level >= lo_bound & pairs$flow_level < hi_bound
  local_n[a] <- sum(in_band)
  if (local_n[a] > 0) {
    w_local <- interpolate_weight_vec(pairs$flow_level[in_band], anchors, rounded_weights)
    ens_local <- (1 - w_local) * pairs$yesterday_lf[in_band] +
                 w_local * pairs$ef_pred[in_band]
    local_rmse[a] <- sqrt(mean((ens_local - pairs$actual[in_band])^2))
  } else {
    local_rmse[a] <- NA
  }
}
out_df <- data.frame(
  anchor_flow    = anchors,
  optimal_weight = rounded_weights,
  n_obs_nearby   = local_n,
  rmse_local     = round(local_rmse, 1)
)
write.csv(out_df, out_path, row.names = FALSE)
cat(sprintf("
Results saved to: %s
", out_path))

# --- 12. Summary -------------------------------------------------------------
cat("
=============================================================
")
cat("  SUMMARY
")
cat("=============================================================
")
cat(sprintf("  Step-function RMSE:       %.1f cfs
", rmse_step))
cat(sprintf("  Gradient (fine) RMSE:     %.4f cfs
", best_rmse))
cat(sprintf("  Gradient (rounded) RMSE:  %.1f cfs
", rmse_rounded))
cat(sprintf("  Improvement (rounded):    %+.1f%% vs step function
",
            (rmse_rounded - rmse_step) / rmse_step * 100))
cat(sprintf("  N observations used:      %d consecutive-day pairs
", nrow(pairs)))
cat(sprintf("  Anchor points:            %d
", n_anchors))
cat(sprintf("  Final weights:            [%s]
",
            paste(sprintf("%.1f", rounded_weights), collapse = ", ")))
cat("=============================================================
")
